Wikipedia/Code/InsertMediaSearchResultsCollectionViewController.swift

```diff
@@ -0,0 +1,261 @@
+}
+
+final class InsertMediaSearchResult {
+    let fileTitle: String
+    let displayTitle: String
+    let thumbnailURL: URL
+    var imageInfo: MWKImageInfo?
+
+    init(fileTitle: String, displayTitle: String, thumbnailURL: URL) {
+        self.fileTitle = fileTitle
+        self.displayTitle = displayTitle
+        self.thumbnailURL = thumbnailURL
+    }
+
+    func imageURL(for width: CGFloat) -> URL? {
+        guard width > 0 else {
+            assertionFailure("width must be greater than 0")
+            return nil
+        }
+        return URL(string: WMFChangeImageSourceURLSizePrefix(thumbnailURL.absoluteString, Int(width))) ?? imageInfo?.canonicalFileURL
+    }
+}
+
+class InsertMediaSearchResultsCollectionViewController: ViewController {
+    private let collectionView: UICollectionView
+    private var flowLayout: FlowLayout {
+        return collectionView.collectionViewLayout as! FlowLayout
+    }
+
+    weak var delegate: InsertMediaSearchResultsCollectionViewControllerDelegate?
+    weak var scrollDelegate: InsertMediaSearchResultsCollectionViewControllerScrollDelegate?
+
+    var originalContentOffsetY: CGFloat = 0
+
+    var searchResults = [InsertMediaSearchResult]() {
+        didSet {
+            assert(Thread.isMainThread)
+            reload()
+        }
+    }
+
+    override init() {
+        collectionView = UICollectionView(frame: .zero, collectionViewLayout: FlowLayout())
+        super.init()
+        collectionView.dataSource = self
+        collectionView.delegate = self
+        scrollView = collectionView
+        title = CommonStrings.searchTitle
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        collectionView.register(InsertMediaSearchResultCollectionViewCell.self, forCellWithReuseIdentifier: InsertMediaSearchResultCollectionViewCell.identifier)
+        view.wmf_addSubviewWithConstraintsToEdges(collectionView)
+        additionalSafeAreaInsets = UIEdgeInsets(top: 0, left: 0, bottom: flowLayout.minimumLineSpacing, right: 0)
+        collectionView.contentInset = UIEdgeInsets(top: 0, left: 12, bottom: 0, right: 12)
+        apply(theme: theme)
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        originalContentOffsetY = collectionView.contentOffset.y
+    }
+
+    func reload() {
+        collectionView.reloadData()
+        updateEmptyState()
+    }
+
+    private func configure(_ cell: InsertMediaSearchResultCollectionViewCell, at indexPath: IndexPath) {
+        let result = searchResults[indexPath.item]
+        cell.configure(imageURL: result.thumbnailURL, imageViewDimension: flowLayout.itemSize.width, caption: result.displayTitle)
```

Comment:
Passing the image view dimension here leads to a mismatch in image size and cell size when rotating on an iPad:
![Simulator Screen Shot - iPad Air 2 - 2019-04-24 at 08 56 56](https://user-images.githubusercontent.com/741327/56661733-5af6ba00-6670-11e9-8e55-6a2735834ec5.png)

Instead of relying on passing that as a configuration parameter and keeping it stored as a property, the `imageDimension` should always match the `size.width` in `sizeThatFits` so it could be calculated at that time:
```
    override func sizeThatFits(_ size: CGSize, apply: Bool) -> CGSize {
        ...
        let imageViewDimension = size.width
        let selectedImageViewDimension = min(35, imageViewDimension * 0.2)
        ...
    }

```

Reply:
I was also seeing an incorrect number of columns on rotate from portrait to landscape - it looks like the invalidation context needed `invalidateFlowLayoutDelegateMetrics`. I committed what I was working with to `spike/flow_layout` - feel free to use or discard

Reply:
Thanks for looking into it! I merged in your branch 👍